### ISSUE TITLE
Cannot install tensorflow-macos

### ISSUE THREAD

**Issue by tetuomi** [tensorflow/tensorflow#51986]
<em>Please make sure that this is a build/installation issue. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:build_template</em>

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Mac mini(M1) ver.12.0(Monterey)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version: 2.5.0
- Python version: 3.9.7
- Installed using virtualenv? pip? conda?: Conda and pip
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**Describe the problem**
I tried to install Tensorflow-metal by following [Getting Started with tensorflow-metal PluggableDevice](https://developer.apple.com/metal/tensorflow-plugin/).

I was successful up to `conda install -c apple tensorflow-deps`,but error occurred in `python -m pip install tensorflow-macos`.

**Provide the exact sequence of commands / steps that you executed before running into the problem**

```
(Download Miniforge3-MacOSX-arm64.sh)
chmod +x ~/Downloads/Miniforge3-MacOSX-arm64.sh
sh ~/Downloads/Miniforge3-MacOSX-arm64.sh
source ~/miniforge3/bin/activate

conda install -c apple tensorflow-deps
python -m pip install tensorflow-macos
python -m pip install tensorflow-metal
```

**Any other info / logs**

error file is [git hub gist](https://gist.github.com/tetuomi/66ff78288c81ff819f947b70746318e2#file-error-txt)

**Comment by tilakrayal** [comment]
@tetuomi ,
Please refer to similar issues #44751,#47782 and for installation please take a look at this [link](https://github.com/apple/tensorflow_macos). It helps.

**Comment by tetuomi** [comment]
@tilakrayal 
Thanks for the response!

I executed bellow command without error.
maybe succeed.
```
pip install --upgrade -t   "${VIRTUAL_ENV}/lib/python3.9/site-packages" --no-dependencies --force tensorflow-macos
```
<br>
<details>
<summary>conda list</summary>

# Name                    Version                   Build  Channel

absl-py                   0.10.0             pyhd8ed1ab_1    conda-forge
aiohttp                   3.7.4.post0      py39h5161555_0    conda-forge
astunparse                1.6.3              pyhd8ed1ab_0    conda-forge
async-timeout             3.0.1                   py_1000    conda-forge
attrs                     21.2.0             pyhd8ed1ab_0    conda-forge
blinker                   1.4                        py_1    conda-forge
brotlipy                  0.7.0           py39h5161555_1001    conda-forge
c-ares                    1.17.2               h3422bc3_0    conda-forge
ca-certificates           2021.5.30            h4653dfc_0    conda-forge
cached-property           1.5.2                hd8ed1ab_1    conda-forge
cached_property           1.5.2              pyha770c72_1    conda-forge
cachetools                4.2.2              pyhd8ed1ab_0    conda-forge
certifi                   2021.5.30        py39h2804cbe_0    conda-forge
cffi                      1.14.6           py39h52b1de0_1    conda-forge
chardet                   4.0.0            py39h2804cbe_1    conda-forge
charset-normalizer        2.0.0              pyhd8ed1ab_0    conda-forge
click                     8.0.1            py39h2804cbe_0    conda-forge
colorama                  0.4.4              pyh9f0ad1d_0    conda-forge
cryptography              3.4.7            py39h73257c9_0    conda-forge
dataclasses               0.8                pyhc8e2a94_3    conda-forge
flatbuffers               2.0.0                hbdafb3b_0    conda-forge
gast                      0.4.0              pyh9f0ad1d_0    conda-forge
google-auth               1.35.0             pyh6c4a22f_0    conda-forge
google-auth-oauthlib      0.4.6              pyhd8ed1ab_0    conda-forge
google-pasta              0.2.0              pyh8c360ce_0    conda-forge
grpcio                    1.38.1           py39h9e1b6db_0    conda-forge
h5py                      3.1.0           nompi_py39h99babb8_100    conda-forge
hdf5                      1.10.6          nompi_h0fc092c_1114    conda-forge
idna                      3.1                pyhd3deb0d_0    conda-forge
importlib-metadata        4.8.1            py39h2804cbe_0    conda-forge
keras                     2.6.0              pyhd8ed1ab_0    conda-forge
keras-preprocessing       1.1.2              pyhd8ed1ab_0    conda-forge
krb5                      1.19.2               hd92b7a7_0    conda-forge
libblas                   3.9.0           11_osxarm64_openblas    conda-forge
libcblas                  3.9.0           11_osxarm64_openblas    conda-forge
libclang                  11.1.0          default_h0fdd720_1    conda-forge
libcurl                   7.79.0               h8fe1914_0    conda-forge
libcxx                    12.0.1               h168391b_0    conda-forge
libedit                   3.1.20191231         hc8eb9b7_2    conda-forge
libev                     4.33                 h642e427_1    conda-forge
libffi                    3.4.2                hbdafb3b_1    conda-forge
libgfortran               5.0.0.dev0      11_0_1_hf114ba7_23    conda-forge
libgfortran5              11.0.1.dev0         hf114ba7_23    conda-forge
liblapack                 3.9.0           11_osxarm64_openblas    conda-forge
libllvm11                 11.1.0               h93073aa_2    conda-forge
libnghttp2                1.43.0               hf3018f0_0    conda-forge
libopenblas               0.3.17          openmp_h5dd58f0_1    conda-forge
libprotobuf               3.18.0               hccf11d3_0    conda-forge
libssh2                   1.9.0                hb80f160_6    conda-forge
llvm-openmp               12.0.1               hf3c4609_1    conda-forge
markdown                  3.3.4              pyhd8ed1ab_0    conda-forge
multidict                 5.1.0            py39h5161555_1    conda-forge
ncurses                   6.2                  h9aa5885_4    conda-forge
numpy                     1.19.5           py39h1f3b974_2    conda-forge
oauthlib                  3.1.1              pyhd8ed1ab_0    conda-forge
openssl                   1.1.1l               h3422bc3_0    conda-forge
opt_einsum                3.3.0              pyhd8ed1ab_1    conda-forge
pip                       21.2.4             pyhd8ed1ab_0    conda-forge
protobuf                  3.18.0           py39hfb83b0d_0    conda-forge
pyasn1                    0.4.8                      py_0    conda-forge
pyasn1-modules            0.2.7                      py_0    conda-forge
pycparser                 2.20               pyh9f0ad1d_2    conda-forge
pyjwt                     2.1.0              pyhd8ed1ab_0    conda-forge
pyopenssl                 20.0.1             pyhd8ed1ab_0    conda-forge
pysocks                   1.7.1            py39h2804cbe_3    conda-forge
python                    3.9.7           h54d631c_1_cpython    conda-forge
python_abi                3.9                      2_cp39    conda-forge
pyu2f                     0.1.5              pyhd8ed1ab_0    conda-forge
readline                  8.1                  hedafd6a_0    conda-forge
requests                  2.26.0             pyhd8ed1ab_0    conda-forge
requests-oauthlib         1.3.0              pyh9f0ad1d_0    conda-forge
rsa                       4.7.2              pyh44b312d_0    conda-forge
scipy                     1.7.0            py39h5060c3b_0    conda-forge
setuptools                58.0.4           py39h2804cbe_0    conda-forge
six                       1.15.0             pyh9f0ad1d_0    conda-forge
sqlite                    3.36.0               h72a2b83_1    conda-forge
tensorboard               2.6.0              pyhd8ed1ab_1    conda-forge
tensorboard-data-server   0.6.0            py39hfb8cd70_0    conda-forge
tensorboard-plugin-wit    1.8.0              pyh44b312d_0    conda-forge
**tensorflow-deps           2.6.0                         0    apple
tensorflow-macos          2.5.0                    pypi_0    pypi
tensorflow-metal          0.1.2                    pypi_0    pypi**
termcolor                 1.1.0                      py_2    conda-forge
tk                        8.6.11               he1e0b03_1    conda-forge
tqdm                      4.62.2             pyhd8ed1ab_0    conda-forge
typing-extensions         3.7.4.3                       0    conda-forge
typing_extensions         3.7.4.3                    py_0    conda-forge
tzdata                    2021a                he74cb21_1    conda-forge
urllib3                   1.26.6             pyhd8ed1ab_0    conda-forge
werkzeug                  2.0.1              pyhd8ed1ab_0    conda-forge
wheel                     0.35.1             pyh9f0ad1d_0    conda-forge
wrapt                     1.12.1           py39h5161555_3    conda-forge
xz                        5.2.5                h642e427_1    conda-forge
yarl                      1.6.3            py39h5161555_2    conda-forge
zipp                      3.5.0              pyhd8ed1ab_0    conda-forge
zlib                      1.2.11            h31e879b_1009    conda-forge

</details>

I was able to use GPU, but test is so slow!
Why?

<details>
<summary>Test code</summary>

```
import tensorflow as tf
#from tensorflow.python.compiler.mlcompute import mlcompute

tf.compat.v1.disable_eager_execution()
#mlcompute.set_mlc_device(device_name='gpu')
#print("is_apple_mlc_enabled %s" % mlcompute.is_apple_mlc_enabled())
#print("is_tf_compiled_with_apple_mlc %s" % mlcompute.is_tf_compiled_with_apple_mlc())
print(f"eagerly? {tf.executing_eagerly()}")
print(tf.config.list_logical_devices())

from tensorflow.keras import datasets, layers, models

(train_images, train_labels), (test_images, test_labels) = datasets.cifar10.load_data()
train_images, test_images = train_images / 255.0, test_images / 255.0
class_names = ['airplane', 'automobile', 'bird', 'cat', 'deer',
               'dog', 'frog', 'horse', 'ship', 'truck']
model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(32, 32, 3)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10))
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])
history = model.fit(train_images, train_labels, epochs=10,
                    validation_data=(test_images, test_labels))
```
</details>


<details>
<summary>Output</summary>

```
Init Plugin
Init Graph Optimizer
Init Kernel
eagerly? False
Metal device set to: Apple M1

systemMemory: 8.00 GB
maxCacheSize: 2.67 GB

2021-09-17 10:31:05.072070: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:305] Could not identify NUMA node of platform GPU ID 0, defaulting to 0. Your kernel may not have been built with NUMA support.
2021-09-17 10:31:05.072453: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:271] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 0 MB memory) -> physical PluggableDevice (device: 0, name: METAL, pci bus id: <undefined>)
[LogicalDevice(name='/device:CPU:0', device_type='CPU'), LogicalDevice(name='/device:GPU:0', device_type='GPU')]
Train on 50000 samples, validate on 10000 samples
2021-09-17 10:31:05.810489: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:305] Could not identify NUMA node of platform GPU ID 0, defaulting to 0. Your kernel may not have been built with NUMA support.
2021-09-17 10:31:05.810523: I tensorflow/core/common_runtime/pluggable_device/pluggable_device_factory.cc:271] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 0 MB memory) -> physical PluggableDevice (device: 0, name: METAL, pci bus id: <undefined>)
2021-09-17 10:31:05.822293: W tensorflow/core/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz
2021-09-17 10:31:05.823214: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:112] Plugin optimizer for device_type GPU is enabled.
2021-09-17 10:31:05.846792: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:112] Plugin optimizer for device_type GPU is enabled.
2021-09-17 10:31:05.911894: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:112] Plugin optimizer for device_type GPU is enabled.
Epoch 1/10
2021-09-17 10:31:05.917312: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:112] Plugin optimizer for device_type GPU is enabled.
50000/50000 [==============================] - ETA: 0s - loss: 1.4909 - accuracy: 0.4600/Users/NAME/.local/lib/python3.9/site-packages/tensorflow/python/keras/engine/training.py:2426: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  warnings.warn('`Model.state_updates` will be removed in a future version. '
2021-09-17 10:31:19.453811: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:112] Plugin optimizer for device_type GPU is enabled.
50000/50000 [==============================] - 15s 300us/sample - loss: 1.4909 - accuracy: 0.4600 - val_loss: 1.2483 - val_accuracy: 0.5509
Epoch 2/10
50000/50000 [==============================] - 14s 283us/sample - loss: 1.1333 - accuracy: 0.6020 - val_loss: 1.0202 - val_accuracy: 0.6373
Epoch 3/10
50000/50000 [==============================] - 14s 284us/sample - loss: 0.9841 - accuracy: 0.6535 - val_loss: 0.9950 - val_accuracy: 0.6497
Epoch 4/10
50000/50000 [==============================] - 14s 285us/sample - loss: 0.8839 - accuracy: 0.6910 - val_loss: 0.8935 - val_accuracy: 0.6936
Epoch 5/10
50000/50000 [==============================] - 14s 284us/sample - loss: 0.8092 - accuracy: 0.7140 - val_loss: 0.8907 - val_accuracy: 0.6943
Epoch 6/10
50000/50000 [==============================] - 14s 284us/sample - loss: 0.7508 - accuracy: 0.7363 - val_loss: 0.8767 - val_accuracy: 0.7036
Epoch 7/10
50000/50000 [==============================] - 14s 283us/sample - loss: 0.6978 - accuracy: 0.7538 - val_loss: 0.8459 - val_accuracy: 0.7169
Epoch 8/10
50000/50000 [==============================] - 14s 283us/sample - loss: 0.6538 - accuracy: 0.7708 - val_loss: 0.8494 - val_accuracy: 0.7113
Epoch 9/10
50000/50000 [==============================] - 14s 284us/sample - loss: 0.6114 - accuracy: 0.7847 - val_loss: 0.8723 - val_accuracy: 0.7148
Epoch 10/10
50000/50000 [==============================] - 14s 282us/sample - loss: 0.5738 - accuracy: 0.7961 - val_loss: 0.8718 - val_accuracy: 0.7182
```
</details>


![image](https://user-images.githubusercontent.com/39522526/133710501-9405e0d6-fd2c-4cc9-9b36-a577457f884e.png)

I wanted to run on CPU, so I uninstalled Tensorflow-metal.(```python -m pip uninstall Tensorflow-metal```)
According to GPU History, did not use GPU.
_Output_ and _Output on CPU_ are not so different.
I think that I don't run properly on CPU.

<details>
<summary>Output on CPU</summary>

eagerly? False
[LogicalDevice(name='/device:CPU:0', device_type='CPU')]
Train on 50000 samples, validate on 10000 samples
2021-09-17 10:53:11.080971: W tensorflow/core/platform/profile_utils/cpu_utils.cc:128] Failed to get CPU frequency: 0 Hz
Epoch 1/10
49984/50000 [============================>.] - ETA: 0s - loss: 2.0294 - accuracy: 0.2540/Users/oamishima/.local/lib/python3.9/site-packages/tensorflow/python/keras/engine/training.py:2426: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  warnings.warn('`Model.state_updates` will be removed in a future version. '
50000/50000 [==============================] - 14s 286us/sample - loss: 2.0293 - accuracy: 0.2540 - val_loss: 1.8248 - val_accuracy: 0.3404
Epoch 2/10
50000/50000 [==============================] - 15s 301us/sample - loss: 1.6499 - accuracy: 0.4058 - val_loss: 1.7012 - val_accuracy: 0.3879
Epoch 3/10
50000/50000 [==============================] - 16s 322us/sample - loss: 1.4730 - accuracy: 0.4674 - val_loss: 1.4277 - val_accuracy: 0.4839
Epoch 4/10
50000/50000 [==============================] - 17s 341us/sample - loss: 1.3658 - accuracy: 0.5095 - val_loss: 1.4735 - val_accuracy: 0.4736
Epoch 5/10
50000/50000 [==============================] - 17s 343us/sample - loss: 1.2800 - accuracy: 0.5431 - val_loss: 1.2749 - val_accuracy: 0.5413
Epoch 6/10
50000/50000 [==============================] - 16s 330us/sample - loss: 1.2105 - accuracy: 0.5713 - val_loss: 1.2096 - val_accuracy: 0.5618
Epoch 7/10
50000/50000 [==============================] - 16s 321us/sample - loss: 1.1456 - accuracy: 0.5956 - val_loss: 1.1682 - val_accuracy: 0.5827
Epoch 8/10
50000/50000 [==============================] - 16s 327us/sample - loss: 1.0915 - accuracy: 0.6166 - val_loss: 1.1136 - val_accuracy: 0.6054
Epoch 9/10
50000/50000 [==============================] - 16s 327us/sample - loss: 1.0422 - accuracy: 0.6350 - val_loss: 1.0747 - val_accuracy: 0.6213
Epoch 10/10
50000/50000 [==============================] - 15s 305us/sample - loss: 0.9947 - accuracy: 0.6507 - val_loss: 1.0297 - val_accuracy: 0.6369
</details>

**Comment by tetuomi** [comment]
I run the above Test code on Google CoLab.
the above output is not slow?
Am I running on GPU properly?

![image](https://user-images.githubusercontent.com/39522526/133716499-a6ee12a1-69d2-41d0-a36d-7d848073531e.png)

**Comment by tilakrayal** [comment]
@tetuomi ,
Glad the suggestion worked for you.Please move this issue to closed status as it answers your above question and for any further queries please open a new issue from [here](https://github.com/tensorflow/models/issues/new/choose).Thanks!

**Comment by tetuomi** [comment]
@tilakrayal 
Thanks for helps!!!
